src/orion/algo/mofa/mofa.py

```diff
@@ -0,0 +1,389 @@
+            self.roi_space,
+            self.threshold,
+            self.cur_n_levels,
+        )
+        self.frozen_param_values.update(frozen_param_values)
+        if (set(self.frozen_param_values.keys()) | set(self.roi_space.keys())) != set(
+            self.space.keys()
+        ):
+            raise RuntimeError(
+                "Some dimensions were lost during selection of new region of interest.\n"
+                f"Frozen parameters: {sorted(self.frozen_param_values.keys())}\n"
+                f"Region of interest: {sorted(self.roi_space.keys())}\n"
+                f"Space: {sorted(self.space.keys())}"
+            )
+
+        if self.is_done:
+            return
+
+        # Create new set of trials
+        self.current_trials_params = self._generate_trials_unfrozen_params(
+            self.roi_space
+        )
+
+        return
+
+    @property
+    def is_done(self):
+        """Return True, if an algorithm holds that there can be no further improvement."""
+        if len(self.frozen_param_values.items()) == len(self.space.items()):
+            return True
+
+        return self.converged or super().is_done
+
+
+def get_factorial_performance_analysis(oa_table, space, n_levels):
+    """Compute the factorial perfomance analysis"""
+    levels = list(range(1, n_levels + 1))
+    factorial_performance_analysis = [[level] for level in levels]
+    for key in space.keys():
+        marginal_means = (
+            oa_table[[key, "objective"]].groupby([key]).mean().reset_index()
+        )
+        dim_levels = set(levels)
+        for _, row in marginal_means.iterrows():
+            dim_level = int(row[key])
+            factorial_performance_analysis[dim_level - 1].append(row["objective"])
+            dim_levels.remove(dim_level)
+
+        for remaining_level in dim_levels:
+            factorial_performance_analysis[remaining_level - 1].append(float("inf"))
+
+    return pd.DataFrame(
+        factorial_performance_analysis, columns=["level"] + list(space.keys())
+    )
+
+
+def get_factorial_importance_analysis(factorial_performance_analysis, space):
+    """Compute the factorial importance analysis"""
+    factorial_importance_analysis = []
+    total_marginal_variance = (
+        np.ma.masked_invalid(
+            factorial_performance_analysis[list(space.keys())].to_numpy()
+        )
+        .var(0)
+        .sum()
+    )
+
+    for key in space.keys():
+        best_level = factorial_performance_analysis[key].argmin() + 1
+        importance = (
+            np.ma.masked_invalid(factorial_performance_analysis[key].to_numpy()).var()
+            / total_marginal_variance
+        )
+        factorial_importance_analysis.append((key, best_level, importance))
+
+    return pd.DataFrame(
+        factorial_importance_analysis, columns=["param", "best_level", "importance"]
+    )
+
+
+def select_new_region_of_interest(
+    factorial_importance_analysis, space, threshold, n_levels
```

Comment:
Nitpick:
```suggestion
    factorial_importance_analysis: dict[str, float], space: Space, threshold: float, n_levels: int
```
Btw add `from __future__ import annotations` as the first line if you want to add this.